### ISSUE TITLE
Gradle demo not working with recent intellij IDEA

### ISSUE THREAD

**Issue by anidotnet** [pf4j/pf4j#195]
I tried the gradle demo with the latest IntelliJ IDEA (2017.3.3) and it is not working anymore. They have recently made some fundamental changes in gradle plugin and it is restricting the gradle project to run properly. It is not finding the extension.idx anymore. Happening for both Java and Kotlin projects. Maven demo are working fine though.

Do you have any insight on this how to avert this?

**Comment by decebals** [comment]
@anidotnet 
Personal, I don't know how can I help in this issue (I don't know/use Gradle).
I feel that you wrote a comment in the past to JetBrains related to this subject (maybe an old version).
Maybe other users that use PF4J and Gradle can help.

**Comment by anidotnet** [comment]
Yes, for reference here is the issue I raised with JB - https://youtrack.jetbrains.com/issue/KT-19823
But unfortunately, pf4j still does not work though they claims they have fixed it. I am desperately looking for a solution to this problem.

**Comment by decebals** [comment]
Why you need the last IDEA? Is it not an option to use an old,  tested, IDEA?

**Comment by anidotnet** [comment]
Yeah, that is also an option, but again I'll miss out on the new features introduced by the IDE :)

**Comment by shirishpandharikar** [comment]
I tried the Gradle demo and it does not work either with Eclipse or Idea. Looking at the Pf4j code DevelopmentPluginClasspath adds the target/classes and target/lib to the plugin classpath. So when the plugin classloader is constructed the plugin classes do not end up within the classloader. My plugins are getting resolved but fail to start with a ClassNotFound exception. Am I doing something wrong here?

With Gradle, the output directories can be customized easily. I tried doing so by adding the following to the plugins/build.gradle file within the subprojects task project.buildDir = 'target'
This fixes the issue of the plugin class not being found. But when I package the plugins using Gradle I end up having 2 extensions.idx files in the META-INF directory.

Any pointers here would be really helpful.

**Comment by decebals** [comment]
@shirishpandharikar Can you show us the PF4J's log?

**Comment by decebals** [comment]
@shirishpandharikar Maybe https://github.com/decebals/pf4j/issues/192 can help.

**Comment by shirishpandharikar** [comment]
Here are the logs from my run:
[main] INFO org.pf4j.demo.Boot - ########################################
[main] INFO org.pf4j.demo.Boot -                PF4J-DEMO                
[main] INFO org.pf4j.demo.Boot - ########################################
[main] INFO org.pf4j.DefaultPluginStatusProvider - Enabled plugins: []
[main] INFO org.pf4j.DefaultPluginStatusProvider - Disabled plugins: [KotlinPlugin]
[main] INFO org.pf4j.DefaultPluginManager - PF4J version 2.0.0 in 'development' mode
[main] INFO org.pf4j.AbstractPluginManager - Plugin '../plugins/plugin3' is disabled
[main] INFO org.pf4j.AbstractPluginManager - Plugin 'welcome-plugin@0.0.1' resolved
[main] INFO org.pf4j.AbstractPluginManager - Plugin 'KotlinPlugin@0.0.1' resolved
[main] INFO org.pf4j.AbstractPluginManager - Plugin 'hello-plugin@0.0.1' resolved
[main] INFO org.pf4j.AbstractPluginManager - Start plugin 'welcome-plugin@0.0.1'
[main] ERROR org.pf4j.DefaultExtensionFactory - org.pf4j.demo.welcome.WelcomePlugin
java.lang.ClassNotFoundException: org.pf4j.demo.welcome.WelcomePlugin
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.pf4j.PluginClassLoader.loadClass(PluginClassLoader.java:115)
	at org.pf4j.DefaultPluginFactory.create(DefaultPluginFactory.java:46)
	at org.pf4j.PluginWrapper.getPlugin(PluginWrapper.java:78)
	at org.pf4j.AbstractPluginManager.startPlugins(AbstractPluginManager.java:329)
	at org.pf4j.demo.Boot.main(Boot.java:56)
Exception in thread "main" java.lang.NullPointerException
	at org.pf4j.AbstractPluginManager.startPlugins(AbstractPluginManager.java:329)
	at org.pf4j.demo.Boot.main(Boot.java:56)

#192 did not help.

**Comment by decebals** [comment]
Please put the PF4J logging level on debug and resend the log.

**Comment by shirishpandharikar** [comment]
Sorry, my bad. Did not add the log4j dependency. Here are the logs with DEBUG enabled

2018-02-01 22:56:58,736 INFO org.pf4j.demo.Boot - ########################################
2018-02-01 22:56:58,737 INFO org.pf4j.demo.Boot -                PF4J-DEMO                
2018-02-01 22:56:58,737 INFO org.pf4j.demo.Boot - ########################################
2018-02-01 22:56:58,768 INFO org.pf4j.DefaultPluginStatusProvider - Enabled plugins: []
2018-02-01 22:56:58,768 INFO org.pf4j.DefaultPluginStatusProvider - Disabled plugins: [KotlinPlugin]
2018-02-01 22:56:58,772 INFO org.pf4j.DefaultPluginManager - PF4J version 2.0.0 in 'development' mode
2018-02-01 22:56:58,772 DEBUG org.pf4j.AbstractPluginManager - Lookup plugins in '../plugins'
2018-02-01 22:56:58,775 DEBUG org.pf4j.AbstractPluginManager - Found 3 possible plugins: [../plugins/plugin1, ../plugins/plugin3, ../plugins/plugin2]
2018-02-01 22:56:58,775 DEBUG org.pf4j.AbstractPluginManager - Use 'org.pf4j.CompoundPluginDescriptorFinder@6659c656' to find plugins descriptors
2018-02-01 22:56:58,775 DEBUG org.pf4j.AbstractPluginManager - Finding plugin descriptor for plugin '../plugins/plugin1'
2018-02-01 22:56:58,775 DEBUG org.pf4j.CompoundPluginDescriptorFinder - 'org.pf4j.PropertiesPluginDescriptorFinder@6d5380c2' is applicable for plugin '../plugins/plugin1'
2018-02-01 22:56:58,775 DEBUG org.pf4j.PropertiesPluginDescriptorFinder - Lookup plugin descriptor in '../plugins/plugin1/plugin.properties'
2018-02-01 22:56:58,786 DEBUG org.pf4j.AbstractPluginManager - Found descriptor PluginDescriptor [pluginId=welcome-plugin, pluginClass=org.pf4j.demo.welcome.WelcomePlugin, version=0.0.1, provider=Decebal Suiu, dependencies=[], description=, requires=*, license=null]
2018-02-01 22:56:58,786 DEBUG org.pf4j.AbstractPluginManager - Class 'org.pf4j.demo.welcome.WelcomePlugin' for plugin '../plugins/plugin1'
2018-02-01 22:56:58,786 DEBUG org.pf4j.AbstractPluginManager - Loading plugin '../plugins/plugin1'
2018-02-01 22:56:58,786 DEBUG org.pf4j.CompoundPluginLoader - 'org.pf4j.DefaultPluginLoader@7591083d' is applicable for plugin '../plugins/plugin1'
2018-02-01 22:56:58,787 DEBUG org.pf4j.AbstractPluginManager - Loaded plugin '../plugins/plugin1' with class loader 'org.pf4j.PluginClassLoader@736e9adb'
2018-02-01 22:56:58,788 DEBUG org.pf4j.AbstractPluginManager - Creating wrapper for plugin '../plugins/plugin1'
2018-02-01 22:56:58,790 DEBUG org.pf4j.AbstractPluginManager - Created wrapper 'PluginWrapper [descriptor=PluginDescriptor [pluginId=welcome-plugin, pluginClass=org.pf4j.demo.welcome.WelcomePlugin, version=0.0.1, provider=Decebal Suiu, dependencies=[], description=, requires=*, license=null], pluginPath=../plugins/plugin1]' for plugin '../plugins/plugin1'
2018-02-01 22:56:58,793 DEBUG org.pf4j.AbstractPluginManager - Use 'org.pf4j.CompoundPluginDescriptorFinder@6659c656' to find plugins descriptors
2018-02-01 22:56:58,794 DEBUG org.pf4j.AbstractPluginManager - Finding plugin descriptor for plugin '../plugins/plugin3'
2018-02-01 22:56:58,794 DEBUG org.pf4j.CompoundPluginDescriptorFinder - 'org.pf4j.PropertiesPluginDescriptorFinder@6d5380c2' is applicable for plugin '../plugins/plugin3'
2018-02-01 22:56:58,794 DEBUG org.pf4j.PropertiesPluginDescriptorFinder - Lookup plugin descriptor in '../plugins/plugin3/plugin.properties'
2018-02-01 22:56:58,795 DEBUG org.pf4j.AbstractPluginManager - Found descriptor PluginDescriptor [pluginId=KotlinPlugin, pluginClass=org.pf4j.demo.kotlin.KotlinPlugin, version=0.0.1, provider=Anindya Chatterjee, dependencies=[], description=, requires=*, license=null]
2018-02-01 22:56:58,795 DEBUG org.pf4j.AbstractPluginManager - Class 'org.pf4j.demo.kotlin.KotlinPlugin' for plugin '../plugins/plugin3'
2018-02-01 22:56:58,795 DEBUG org.pf4j.AbstractPluginManager - Loading plugin '../plugins/plugin3'
2018-02-01 22:56:58,796 DEBUG org.pf4j.CompoundPluginLoader - 'org.pf4j.DefaultPluginLoader@7591083d' is applicable for plugin '../plugins/plugin3'
2018-02-01 22:56:58,796 DEBUG org.pf4j.AbstractPluginManager - Loaded plugin '../plugins/plugin3' with class loader 'org.pf4j.PluginClassLoader@2d38eb89'
2018-02-01 22:56:58,796 DEBUG org.pf4j.AbstractPluginManager - Creating wrapper for plugin '../plugins/plugin3'
2018-02-01 22:56:58,797 INFO org.pf4j.AbstractPluginManager - Plugin '../plugins/plugin3' is disabled
2018-02-01 22:56:58,797 DEBUG org.pf4j.AbstractPluginManager - Created wrapper 'PluginWrapper [descriptor=PluginDescriptor [pluginId=KotlinPlugin, pluginClass=org.pf4j.demo.kotlin.KotlinPlugin, version=0.0.1, provider=Anindya Chatterjee, dependencies=[], description=, requires=*, license=null], pluginPath=../plugins/plugin3]' for plugin '../plugins/plugin3'
2018-02-01 22:56:58,798 DEBUG org.pf4j.AbstractPluginManager - Use 'org.pf4j.CompoundPluginDescriptorFinder@6659c656' to find plugins descriptors
2018-02-01 22:56:58,798 DEBUG org.pf4j.AbstractPluginManager - Finding plugin descriptor for plugin '../plugins/plugin2'
2018-02-01 22:56:58,798 DEBUG org.pf4j.CompoundPluginDescriptorFinder - 'org.pf4j.PropertiesPluginDescriptorFinder@6d5380c2' is applicable for plugin '../plugins/plugin2'
2018-02-01 22:56:58,798 DEBUG org.pf4j.PropertiesPluginDescriptorFinder - Lookup plugin descriptor in '../plugins/plugin2/plugin.properties'
2018-02-01 22:56:58,800 DEBUG org.pf4j.AbstractPluginManager - Found descriptor PluginDescriptor [pluginId=hello-plugin, pluginClass=org.pf4j.demo.hello.HelloPlugin, version=0.0.1, provider=Decebal Suiu, dependencies=[], description=, requires=*, license=null]
2018-02-01 22:56:58,800 DEBUG org.pf4j.AbstractPluginManager - Class 'org.pf4j.demo.hello.HelloPlugin' for plugin '../plugins/plugin2'
2018-02-01 22:56:58,800 DEBUG org.pf4j.AbstractPluginManager - Loading plugin '../plugins/plugin2'
2018-02-01 22:56:58,800 DEBUG org.pf4j.CompoundPluginLoader - 'org.pf4j.DefaultPluginLoader@7591083d' is applicable for plugin '../plugins/plugin2'
2018-02-01 22:56:58,811 DEBUG org.pf4j.AbstractPluginManager - Loaded plugin '../plugins/plugin2' with class loader 'org.pf4j.PluginClassLoader@5fa7e7ff'
2018-02-01 22:56:58,811 DEBUG org.pf4j.AbstractPluginManager - Creating wrapper for plugin '../plugins/plugin2'
2018-02-01 22:56:58,812 DEBUG org.pf4j.AbstractPluginManager - Created wrapper 'PluginWrapper [descriptor=PluginDescriptor [pluginId=hello-plugin, pluginClass=org.pf4j.demo.hello.HelloPlugin, version=0.0.1, provider=Decebal Suiu, dependencies=[], description=, requires=*, license=null], pluginPath=../plugins/plugin2]' for plugin '../plugins/plugin2'
2018-02-01 22:56:58,817 DEBUG org.pf4j.DependencyResolver - Graph: 
   welcome-plugin -> []
   KotlinPlugin -> []
   hello-plugin -> []
2018-02-01 22:56:58,818 DEBUG org.pf4j.DependencyResolver - Plugins order: [welcome-plugin, KotlinPlugin, hello-plugin]
2018-02-01 22:56:58,820 INFO org.pf4j.AbstractPluginManager - Plugin 'welcome-plugin@0.0.1' resolved
2018-02-01 22:56:58,821 INFO org.pf4j.AbstractPluginManager - Plugin 'KotlinPlugin@0.0.1' resolved
2018-02-01 22:56:58,821 INFO org.pf4j.AbstractPluginManager - Plugin 'hello-plugin@0.0.1' resolved
2018-02-01 22:56:58,821 INFO org.pf4j.AbstractPluginManager - Start plugin 'welcome-plugin@0.0.1'
2018-02-01 22:56:58,821 DEBUG org.pf4j.DefaultExtensionFactory - Create instance for plugin 'org.pf4j.demo.welcome.WelcomePlugin'
2018-02-01 22:56:58,826 ERROR org.pf4j.DefaultExtensionFactory - org.pf4j.demo.welcome.WelcomePlugin
java.lang.ClassNotFoundException: org.pf4j.demo.welcome.WelcomePlugin
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.pf4j.PluginClassLoader.loadClass(PluginClassLoader.java:115)
	at org.pf4j.DefaultPluginFactory.create(DefaultPluginFactory.java:46)
	at org.pf4j.PluginWrapper.getPlugin(PluginWrapper.java:78)
	at org.pf4j.AbstractPluginManager.startPlugins(AbstractPluginManager.java:329)
	at org.pf4j.demo.Boot.main(Boot.java:56)
Exception in thread "main" java.lang.NullPointerException
	at org.pf4j.AbstractPluginManager.startPlugins(AbstractPluginManager.java:329)
	at org.pf4j.demo.Boot.main(Boot.java:56)

**Comment by decebals** [comment]
Please put the logging level for PluginClassLoader on TRACE and add the log.

**Comment by shirishpandharikar** [comment]
These are the TRACE logs for the PluginClassLoader

2018-02-02 13:18:26,934 TRACE org.pf4j.PluginClassLoader - Received request to load class 'org.pf4j.demo.welcome.WelcomePlugin'
2018-02-02 13:18:26,934 TRACE org.pf4j.PluginClassLoader - Received request to load class 'org.pf4j.demo.welcome.WelcomePlugin'
2018-02-02 13:18:26,934 TRACE org.pf4j.PluginClassLoader - Delegate the loading of class 'org.pf4j.demo.welcome.WelcomePlugin' to parent
2018-02-02 13:18:26,934 TRACE org.pf4j.PluginClassLoader - Delegate the loading of class 'org.pf4j.demo.welcome.WelcomePlugin' to parent
2018-02-02 13:18:26,936 TRACE org.pf4j.PluginClassLoader - Search in dependencies for class 'org.pf4j.demo.welcome.WelcomePlugin'
2018-02-02 13:18:26,936 TRACE org.pf4j.PluginClassLoader - Search in dependencies for class 'org.pf4j.demo.welcome.WelcomePlugin'
2018-02-02 13:18:26,936 TRACE org.pf4j.PluginClassLoader - Couldn't find class 'org.pf4j.demo.welcome.WelcomePlugin' in plugin classpath. Delegating to parent
2018-02-02 13:18:26,936 TRACE org.pf4j.PluginClassLoader - Couldn't find class 'org.pf4j.demo.welcome.WelcomePlugin' in plugin classpath. Delegating to parent

**Comment by decebals** [comment]
@shirishpandharikar 
Not only TRACE entries. They are other lines on DEBUG level (https://github.com/decebals/pf4j/blob/master/pf4j/src/main/java/org/pf4j/PluginClassLoader.java#L52)

**Comment by shirishpandharikar** [comment]
The code at [PluginClassLoader.java#L52](https://github.com/decebals/pf4j/blob/master/pf4j/src/main/java/org/pf4j/PluginClassLoader.java#L52) never gets executed. The reason is the DefaultPluginClasspath uses classesDirectories as **target/classes**. Please refer to [DefaultPluginLoader.java#L67](https://github.com/decebals/pf4j/blob/master/pf4j/src/main/java/org/pf4j/DefaultPluginLoader.java#L67) which checks if the folder/file exists. In case of Gradle the classes are generated in a different location e.g. **build/classes/java/main**. 

I ended up extending the DevelopmentPluginClasspath class and overriding the `createPluginClasspath` method in DefaultPluginManager class to use this new class.

```java
private static class GradleDevelopmentPluginClasspath extends DevelopmentPluginClasspath {
    public GradleDevelopmentPluginClasspath() {
        super();
        addClassesDirectories("build/classes/java/main", "build/resources/main");            
    }
}
```
```java
@Bean
public SpringPluginManager pluginManager() {
    return new SpringPluginManager() {
    @Override
    protected PluginClasspath createPluginClasspath() {
        return isDevelopment() ? new GradleDevelopmentPluginClasspath() : new DefaultPluginClasspath();
     }
  };
}
```

**Comment by decebals** [comment]
@shirishpandharikar 
Your approach is correct. 
I don't know if help to add a logging line (maybe with warn or debug level) in [DefaultPluginLoader.java#L67](https://github.com/decebals/pf4j/blob/master/pf4j/src/main/java/org/pf4j/DefaultPluginLoader.java#L67) in a new else block, maybe it will be useful for debug.

**Comment by ssherwood** [comment]
This issue will also impact development mode for Gradle + Kotlin projects.  I recently stumbled across this myself and had to do something similar to @shirishpandharikar (only specific to Kotlin outputs):

- build/classes/kotlin/main
- build/tmp/kapt3/classes/main

I'm not sure there is a straightforward way to account for all the possible classpath variations that might occur with a Gradle project.

**Comment by decebals** [comment]
I don't use Gradle, Kotlin so I cannot help more. If you have some suggestions about how to improve this aspect please let me know. From what I understand, the `development` mode ([DevelopmentPluginClasspath](https://github.com/pf4j/pf4j/blob/master/pf4j/src/main/java/org/pf4j/DevelopmentPluginClasspath.java) with `target/classes` and `target/lib`) works out of the box only for Maven projects.

**Comment by ssherwood** [comment]
I don't have a great answer, but I'll run through a couple of thoughts that I have (in no specific order):

- Add the additional likely paths to the existing DevelopmentPluginClasspath.  This would mix Maven and Gradle build outputs together in the same config but it may not negatively impact anyone (i.e. you're either one or the other).  It might create issues since there are additional paths that will be checked.

- Add additional development modes.  This would add new values to the RuntimeMode enum for the other variations (DEVELOPMENT_GRADLE, DEVELOPMENT_KOTLIN, etc.).  This might create too many variations that will just end up confusing people.

- Make the existing DevelopmentPluginClasspath more easily configurable w/o needing to override the entire DefaultPluginManager.  This still may not be obvious enough to developers that aren't using Maven and still cause confusion when using development mode.

- Create a Maven/Gradle plugin that is smart enough to generate the correct metadata/config that can be used by the plugin run-time to determine which paths to evaluate.  This could get really complicated and be a lot of extra work for limited value.

- Do nothing, just update the documentation for Gradle/Kotlin developers to make it clear that they have to override the plugin manager to get development mode working.

Off the top of my head, these are all the ideas that I could come up with.  I can't say I like any one specifically better than the other but I would lean towards something more than just the documentation approach.

**Comment by decebals** [comment]
Do you think that it's a good idea if we split `DevelopmentPluginClasspath` in:
- `DevelopmentPluginClasspath.Maven` (`classesDirectories` = "target/classes", `libDirectories` = "target/lib)
- `DevelopmentPluginClasspath.Gradle` (`classesDirectories` = ?, `libDirectories` = "?)

What values we have for `classesDirectories`  and `libDirectories` in case of Gradle?

Do we need also a `DevelopmentPluginClasspath.Kotlin`? If the response is yes, what values we have for `classesDirectories`  and `libDirectories` in this case?

**Comment by ssherwood** [comment]
For the classesDirectory of Gradle + Java, the folders are (per @shirishpandharikar examples):
- build/classes/java/main
- build/resources/main

For the classesDirectory of Gradle + Kotlin (without Kapt), they are:
- build/classes/kotlin/main
- build/resources/main

However, if you use Kapt to create the extensions.idx via annotation scanning, you need to add:
- build/tmp/kapt3/classes/main

In theory, for the DevelopmentPluginClasspath.Gradle, you can just add all of these together and if the folder exists, its part of path (otherwise I don't think it is creating an error condition).

For the libDirectories, I'm actually not sure how they apply in the Development mode.  Mine are currently still pointing to a non-existent Maven folder and I'm not seeing any issues.  Is this config actually used in this mode?  Or is it just working for me because my IDE (IntelliJ) is providing them in the classpath already?

**Comment by decebals** [comment]
I propose you an improvement (PR #219). Please take a look, test (I tested with Maven) and tell me your opinion.

**Comment by ssherwood** [comment]
This appears to works great in my initial testing.  My gut feel is that it will cover the majority of scenarios for most projects (and at this point it is easy enough to override and extend if needed).  Thanks for the quick turn around.

**Comment by decebals** [comment]
Can we close this issue?

**Comment by ssherwood** [comment]
Sorry, I hate to be a pest but I just realized that I've been working around one other little annoyance when using Gradle (sigh).  When in development mode, if the plugins folder has a "build" output directory, I get an error in the logs "Field 'id' cannot be empty" (it was looking in the generated build output folder for a plugin).

I realized that there is a little work around built-in for Maven projects that skip the "target" folder in development mode:

https://github.com/pf4j/pf4j/blob/7eeea8e7f01b39811de20280f5e9ba51a456252a/pf4j/src/main/java/org/pf4j/DefaultPluginRepository.java#L78

Adding the default Gradle output "build" folder to this list would eliminate the need for my work around (and save a future Gradle user from scratching their head over this one).
